proposals/umoci.md

```diff
@@ -0,0 +1,113 @@
+
+* Aleksa Sarai <cyphar@cyphar.com> (@cyphar)
+* Tycho Andersen <tycho@tycho.ws> (@tych0)
+* Vincent Batts <vbatts@hashbangbash.com> (@vbatts)
+
+### Code of Conduct ###
+This project would incorporate (by reference) the OCI [Code of Conduct][code-of-conduct].
+
+[code-of-conduct]: https://github.com/opencontainers/org/blob/master/CODE_OF_CONDUCT.md
+
+### Governance and Releases ###
+This project would incorporate the Governance and Releases processes from the OCI project template: https://github.com/opencontainers/project-template.
+
+It should be noted that since umoci is not a specification, it is not bound by the ordinary quorum and voting rules for specification release.
+As such, new versions will be released as regularly as needed without the need for a quorum vote.
+This is to avoid the serious delays in releases encountered within the runc project.
+
+Until there are enough additional maintainers, PRs will be merged with only one maintainer LGTM required.
+Maintainers are encouraged to not merge their own PRs immediately (preferably waiting at least 24 to 48 hours to allow another maintainer to review it), but this is not a strict rule.
+
+### Project Communications ###
+The proposed project would continue to use existing channels in use by the OCI developer community for communication including:
+
+ * GitHub for issues and pull requests.
+ * The [`dev@opencontainers.org`][oci-ml] email list.
+ * The weekly OCI developer community conference call.
+ * The `#opencontainers` IRC channel on Freenode.
+ * The [OCI Slack workspace][oci-slack].
+ * The [OCI Matrix Room][oci-matrix].
+
+[oci-ml]: mailto:dev@opencontainers.org
+[oci-slack]: https://opencontainers.slack.com/
+[oci-matrix]: https://matrix.to/#/#opencontainers:matrix.org
+
+### Roadmap ###
+umoci still has a fairly large scope of work, and this work would continue under OCI.
+Examples of such work include:
+
+ * Hookable execution to allow for custom storage drivers to be defined by users (to avoid the maintenance overhead of supporting such drivers within umoci).
+ * Security hardening by porting umoci to [libpathrs][libpathrs].
+ * Re-thinking the method by which images are referenced in the command-line.
+ * Expanded support for configuring descriptor annotations in arbitrarily-structured OCI metadata trees.
+ * Committing to a stable Go API and releasing a 1.0 of the project.
+ * Experimental design work for a new OCI image-spec layer format (code-named "OCIv2").
+
+[libpathrs]: https://github.com/openSUSE/libpathrs
+
+## Frequently Asked Questions (FAQ)
+> *Does this proposal change the OCI Charter?*
+
+This proposal does not intend to amend the [OCI Charter][oci-charter].
+
+[oci-charter]: https://github.com/opencontainers/tob/blob/master/CHARTER.md
+
+> *Where does umoci fit into the OCI suite of projects?*
+
+umoci is intended to be a reference implementation of the OCI image-spec, and to fill a role akin to runc's relationship with the OCI runtime-spec.
```

Comment:
> I like the umoci web description ... "tool for manipulating and interacting with container images" compatible with the OCI image-spec.

Well, I can't really call it a "reference implementation" unless it's accepted into the OCI as such.

> Why should existing tools drop their code and go with a runc like cli model based on umoci image services?

Being a reference implementation doesn't mean that users must use it or switch to it, it simply means that if a user wishes to make use of a specification there is a clear starting point for doing so.

`runc`'s role as a reference implementation is separate from its role as a de-facto standard container runtime which almost everyone uses. The latter is both historic as well as the result of a series of missteps on the part of runc and runtime-spec developers. However, even in `runc`'s case there are several other runtime-spec implementations that are seeing fairly wide use (such as `crun` for instance).

Reply:
right and most container runtimes have the option of configuring to use either runc, crun, or other runc compatible, e.g. kata. This because at the beginning there was runc, and now even kubernetes has a runtime switch via CRI. Referring to it in this way, as a reference implementation begs the question I think.

Reply:
The runtime-spec does mandate certain operations and a whole container lifecylce though, so it's not really the same as the image-spec in terms of what a "reference implementation" would entail. `umoci` lets you unpack, repack, and configure OCI image-spec images -- and I'm not really sure what else a "reference implementation" of the OCI image-spec could be.

And since the image-spec doesn't mandate any operations (it's purely an on-disk format), I don't think being able to plug out different OCI image-spec implementations is something that makes sense. The point of interoperability for the image-spec is the resulting image format, not what features the program creating it has. This is in stark contrast to the runtime-spec, where interoperability necessarily requires you to be able to swap out the container runtime.

Reply:
I think the comparison to runc isn't helping clarify much here, because while I think the comparison makes sense (it's a "boring container infrastructure" tool which can be used to build larger systems, much like runc) it seems like other folks like yourself put more emphasis on the fact it is ostensibly a pluggable piece of "boring container infrastructure". As I said above, that pluggability is an aspect of the runtime-spec not of runc itself, but I can see how the comparison may be muddying the waters. I'll take a shot a rephrasing it.

Reply:
right, I think we are in violent agreement :-)